### ISSUE TITLE
Proposal: from current-gen References / Definitions / Type Definitions / Implementations etc -> to "arbitrary custom, language-server-provided named lookups"

### ISSUE THREAD

**Issue by metaleap** [microsoft/language-server-protocol#919]
The various "predefined Location queries / lookups" in the LSP such as _References_, _Definitions_, _Type Definitions_, _Implementations_ etc. all tend to show up in clients as usually outstanding UX for browsing a-list-of source-code-locations (uri+pos).

Now while these predefined lookups are the most common-denominator ones applicable to the vast majority of languages out there, they're (depending on language-specific peculiarities / semantics) almost never the _only_ ones users would want and could leverage.

Example: some 2 years back I implemented a _VS Code_ UX capturing the various Golang lookups that the [guru](https://godoc.org/golang.org/x/tools/cmd/guru) source-intel program exposes:

![2020-02-14](https://user-images.githubusercontent.com/22109/74573700-a40e5680-4f81-11ea-8da2-6c2077f043c7.png)

**You guessed it:** almost all these search functionalities offered by `guru` yield a list of path+pos locations! But alas, those _results listings_ themselves had to be rendered into plain old QuickPick dropdowns to open the corresponding resulting Location's URI and jump to the resulting Location's Position.

Now that I'm embarking on a Zig Language Server, I'm again already coming up with numerous scenarios to be able to offer especially VS Code users the reuse of the brilliant Peek UX here. (And these ideas are truly Zig-specific, not something specifically to propose to LSP to introduce as a new "predefined lookup" in the vein of References, Definitions etc.)

**We know this: all the current LSP-predefined lookups boil down to producing a list of source-file `Location`s / `LocationLink`s.** I propose a new and simple ServerCapability to easily expose any number of additional such lookups at init-time (each being merely some free-form id and a user-facing name), plus a new `Request` similar in shape to the existing References / Definitions / Implementations etc. except also taking in its `FooParams` a custom-lookup's identifier that was announced in `ServerCapabilities`.

As for clients? They could choose whatever manner to offer these Language-Server-provided "custom lookups" in their UI but for the example of VS Code, might put them in the editor context menu's `Peek` sub-menu. I'm supposing a corresponding `boolean` ClientCapability might also be introduced here, but overall I see this as a comparatively trouble-and-friction-free addition for language-client developers, given they already have all the machinery for their "lookup / peek UX" in place, including the taking of input `Location[]`s or `LocationLink[]`s to populate such.

**Comment by dbaeumer** [comment]
I agree that for general navigation a Location / LocationLink result is sufficient. However from the experience of implementing call hierarchy support (https://github.com/microsoft/vscode-languageserver-node/blob/master/protocol/src/protocol.callHierarchy.proposed.ts#L21) there are things that are more complicated than this.

From a UI perspective I think we at lease need to know whether the request is a goto or a find request since editors might have more custom UI for this.

**Comment by jedwards1211** [comment]
React devs might want to do a React render hierarchy which would have a very similar response structure/UI to call hierarchy, however definitely should be a separate action from "open call hierarchy".  So I think @metaleap has a very valid point.
